### ISSUE TITLE
Extension Test API  - run results are not correctly updated when using multiple vscode.TestRun objects

### ISSUE THREAD

**Issue by vscodenpa** [vscodenpa/testissues#21512]
Does this issue occur when all extensions are disabled?: Yes (with the exception the extension code in question)

- VS Code Version:  1.77.2 (but this bug is not new)
- OS Version: Ubuntu 22.04

Imagine you have a multiroot workpace containing 3 project folders.
As an extension author, you want to create a `vscode.TestRun` object for each project folder, and kick off all three test runs asynchronously (i.e. in parallel) from a single `vscode.TestRunRequest`.

Unfortunately, this results in:
- `run.passed` not being honoured by the UI until the end of a different test run (or potentially all runs)
- `run.skipped` not being honoured at all (showing as not run rather than skipped).
(run.failed works correctly)

Given the above, if the code sequence is:
`run.enqueued(test) -> run.started(test) -> run.passed(test) -> run.skipped(test2) -> run.end()`
Then what the user will actually see in the UI is:
`run.enqueued(test) -> run.started(test) -> run.enqueued(test) -> run.end() -> run.passed(test) ` 
( run.skipped(test2) has no UI effect and the test is shown as not run)

i.e. the test items are stuck in a run.enqueued state until the end of a run (except for `run.failed`)
What is stranger is that it's not the run that is ended that has its tests updated, but a _different_ run.
The run objects seem to be _interacting with each other_, i.e. they are not isolated.
This lack of isolation between run objects is confusing as an extension author (and most likely unintended, as it is not mentioned in documentation).
Equally, `run.appendOutput` also uses an output terminal shared between all run objects and therefore if all tests are run, then the output from all project folders (likely with different test runners) is intertwined and not isolated.

Moving on to another example, imagine there are two test extensions from different authors in the same workspace.
These extensions create their own `vscode.TestRun` objects, but what happens is:
- tests are not fully updated until the _other_ extension ends its run
- `run.appendOutput` of the two run objects (and therefore two extensions) is intertwined in the test output terminal window (and testresults view), consequently any test output is not human-readable, such that the extension authors must use an extension-specific output window instead to isolate their extension test output (especially in the case of BDD output). or they could buffer the output, but buffering is not a good experience for the user. (guidance on the _intended_ MS design to avoid intertwined test output between extensions would be appreciated, given separate run objects are not sufficient to isolate the output.)

The extension author may never realise these - exist during their testing unless they are testing with another test extension installed.
The manifestation of this problem with two Microsoft extensions has previously been observed and raised [here](https://github-com/microsoft/vscode/-/149328-issuecomment-1125285858).

Here is a gif of the behaviour using a single extension with three run objects, one per project:

![GIF](https://raw.githubusercontent.com/jimasp/test-provider-sample_runobjects-example/main/sample-1681555083759.gif)

Below is the code that produced the above gif. 
Note that is intentionally not a mulitroot workspace, just a simulation of one. 
This is to show that the problem is not limited/caused by multiroot workspaces (or multiple extensions) but simply because run objects are not isolated from each other.

```
// extension.ts

import * as vscode from 'vscode';

const projectFolders = ['project1', 'project2', 'project3'];

export function activate(context: vscode.ExtensionContext) {
	const ctrl = vscode.tests.createTestController('runtests', 'runtests');
	context.subscriptions.push(ctrl);
	const runHandler = testRunHandler(ctrl);

	// simulate multiroot workspace with 3 project folders
	for (const name of projectFolders) {
		const projectItem = ctrl.createTestItem(name, name);
		ctrl.items.add(projectItem);		
		createTestsForProjectFolder(ctrl, projectItem);
	}


	ctrl.createRunProfile('Run Tests', vscode.TestRunProfileKind.Run,
		async (request: vscode.TestRunRequest) => {
			await vscode.commands.executeCommand("testing.clearTestResults");			
			await runHandler(request);
		});
}

function createTestsForProjectFolder(ctrl: vscode.TestController, parent: vscode.TestItem) {
	const outcomes = ["pass", "fail", "skip"];
	for (const outcome of outcomes) {
		const id = `${parent.label} ${outcome}`;
		const test = ctrl.createTestItem(id, id);
		ctrl.items.add(test);
		parent.children.add(test);
	}
}

function testRunHandler(ctrl: vscode.TestController) {
	// kick off each project folder test run async (i.e. in parallel)
	return async (request: vscode.TestRunRequest) => {
		for (const name of projectFolders) {
			const run = ctrl.createTestRun(request, name);
			const projectQueue = getProjectQueue(ctrl, name);
			ctrl.items.forEach(test => {if(test.id.startsWith(name + " ")) projectQueue.push(test);});
			projectQueue.forEach(test => run.enqueued(test));
			runProjectQueueAsync(run, projectQueue);
		}		
	};
}

async function runProjectQueueAsync(run:vscode.TestRun, projectQueue:vscode.TestItem[]) {
	for(const item of projectQueue) {
		// run queue sequentially
		if(run.token.isCancellationRequested)
			break;
		run.started(item);			
		await new Promise(r => setTimeout(r, Math.random() * 5000));
		switch(item.label.split(" ")[1]) {
			case "pass": run.passed(item); break;
			case "fail": run.failed(item, new vscode.TestMessage("failed")); break;
			case "skip": run.skipped(item); break;
		}
		run.appendOutput(item.label + " result updated\r\n");
	}
	await vscode.window.showInformationMessage(`all tests updated for ${run.name}. click OK to end run ${run.name}`, "OK");	
	run.end();
}

function getProjectQueue(ctrl:vscode.TestController, projectFolerName:string) {
	const pqItem = ctrl.items.get(projectFolerName);
	const queue: vscode.TestItem[] = [];
	pqItem?.children.forEach(c => queue.push(c));
	return queue;
}
```

Full repo for the above example is [here](https://github-com/jimasp/test-provider-sample_runobjects-example). 
But note that the repo is basically a duplicate of the MS "test-provider-sample" except that `extension.ts` is replaced with the above code, so alternatively you can just copy/paste the above code into that file if you already have the MS repo locally.
